File: 2016/08.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;
#define PRINTLN(message) cout << message << endl;
#define PRINT(message) cout << message;

const int HEIGHT = 6;
const int WIDTH = 50;

char map[HEIGHT][WIDTH];

vector<string> split(const string &str, char c) {
    vector<string> strings;
    int index = str.find(c);
    strings.push_back(str.substr(0, index));
    strings.push_back(str.substr(index + 1));
    return strings;
}

void create_rect(vector<string> &command) {
    auto strings = split(command[1], 'x');
    int xx = stoi(strings[0]);
    int yy = stoi(strings[1]);

    for (int y = 0; y < yy; ++y) {
        for (int x = 0; x < xx; ++x) {
            map[y][x] = '#';
        }
    }
}

void rotate_column(vector<string> &command) {
    auto strings = split(command[2], '=');
    int where = stoi(strings[1]);
    int by = stoi(command[4]);

    vector<char> chars;
    for (int y = 0; y < HEIGHT; y++) {
        chars.emplace_back(map[(y - by + HEIGHT) % HEIGHT][where]);
    }

    for (int y = 0; y < HEIGHT; y++) {
        map[y][where] = chars[y];
    }
}

void rotate_row(vector<string> &command) {
    auto strings = split(command[2], '=');
    int where = stoi(strings[1]);
    int by = stoi(command[4]);

    vector<char> chars;
    for (int x = 0; x < WIDTH; x++) {
        chars.emplace_back(map[where][(x - by + WIDTH) % WIDTH]);
    }

    for (int x = 0; x < WIDTH; x++) {
        map[where][x] = chars[x];
    }
}

int main() {
    ifstream input("08.in");
    vector<vector<string>> commands;

    for (string line; getline(input, line);) {
        istringstream iss(line);
        vector<string> command;
        for (string word; getline(iss, word, ' ');) {
            command.push_back(word);
        }
        commands.push_back(command);
    }

    for (int y = 0; y < HEIGHT; ++y) {
        for (int x = 0; x < WIDTH; ++x) {
            map[y][x] = ' ';
        }
    }

    for (vector<string> &command : commands) {
        if (command[0] == "rect")
            create_rect(command);
        else if (command[1] == "column")
            rotate_column(command);
        else
            rotate_row(command);
    }

    int count = 0;

    PRINTLN();
    for (int y = 0; y < HEIGHT; ++y) {
        for (int x = 0; x < WIDTH; ++x) {
            if (map[y][x] == '#') {
                ++count;
            }
            PRINT(map[y][x]);
        }
        PRINTLN();
    }
    PRINTLN();

    PRINTLN(count);
}

File: 2016/08.py
commands = [line.strip() for line in open("08.in").readlines()]

HEIGHT = 6
WIDTH = 50
MAP = [[' ' for x in range(WIDTH)] for y in range(HEIGHT)]

for command in commands:
    command = command.split()

    if command[0] == "rect":
        xx, yy = command[1].split("x")
        for y in range(int(yy)):
            for x in range(int(xx)):
                MAP[y][x] = "#"

    elif command[0] == "rotate":
        where = int(command[2].split("=")[1])
        by = int(command[4])

        if command[1] == "column":
            new_column = []
            for y in range(HEIGHT):
                new_column.append(MAP[(y-by) % HEIGHT][where])
            for y in range(HEIGHT):
                MAP[y][where] = new_column[y]
        else:
            new_row = []
            for x in range(WIDTH):
                new_row.append(MAP[where][(x-by) % WIDTH])
            for x in range(WIDTH):
                MAP[where][x] = new_row[x]

count = 0

for y in range(HEIGHT):
    for x in range(WIDTH):
        print(MAP[y][x], end="")
        if MAP[y][x] == "#":
            count += 1
    print()

print(count)